### PR TITLE
feat(warning-page): Implement new design and action on opt-in button click

### PR DESCRIPTION
**Pull request by divyanshiGupta**
Parent Story: https://openshift.io/openshiftio/Openshift_io/plan/detail/1110

This PR adds : 

- New design of the warning page.
- Service for enabling feature when user opts-in by clicking on the opt-in button.
- Option in the `feature-wrapper` to either show the default warning page with the option to opt-in the feature or a different template provided by the parent component when the feature is not enabled.

Note : 
I have tested the above changes with the ngx-feature-flag demo app. I also tried to test these changes with fabric8-ui and planner by linking but I am getting `Null Injector for HttpClient` error. This might be due to something going wrong while linking or because of the way the dependencies are being resolved in the new monorepo structure. 
I think the linking issues will be resolved once we move ngx-feature-flag to monorepo.
@christianvogt what do you think about it?
feature-toggle

### CODE REVIEW

**Comment by divyanshiGupta** [comment]
@rohitkrai03 please review the PR

File: projects/ngx-feature-flag/src/lib/feature-wrapper/feature-toggle.component.ts [hunk 1]
```diff
@@ -1,32 +1,51 @@
...
   ngOnInit(): void {
     if (!this.featureName) {
       throw new Error('Attribute `featureName` should not be null or empty');
     }
-
-    this.enabled = this.featureService.isFeatureUserEnabled(this.featureName);
+    this.feature$ = this.featureService.getFeature(this.featureName).pipe(
+      map((feature: Feature) => {
+        if (feature.attributes) {
+          return {
```
Comment (rohitkrai03) [inline projects/ngx-feature-flag/src/lib/feature-wrapper/feature-toggle.component.ts]: Type of `feature$` is defined as `Observable<{} | Feature>`, but here you are mapping the `Feature` type into `{enabled: boolean, level: string}`. So, `feature$` will never have a type `Observable<Feature>`.

File: projects/ngx-feature-flag/src/lib/feature-wrapper/feature-toggle.component.ts [hunk 2]
```diff
@@ -1,32 +1,51 @@
...
+            enabled: feature.attributes.enabled && feature.attributes['user-enabled'],
+            level: feature.attributes['enablement-level']
+          };
+        } else {
+          return {};
+        }
+      }),
+      tap((feature: any) => {
```
Comment (rohitkrai03) [inline projects/ngx-feature-flag/src/lib/feature-wrapper/feature-toggle.component.ts]: Better add a type interface for it instead of using `any`.

File: projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';
```
Comment (rohitkrai03) [inline projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts:1]: @divyanshiGupta I guess this is a duplicate service from `fabric8-ui`. I am thinking if there is a way to reuse the same service here but because of cyclic dependency it would not be easy. We could have it done the way `PipelineService` is done in `ngx-launcher` but thats not the best way for me as well. I think this will be solved by the SDK refactoring.

File: projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.html [hunk 4]
```diff
@@ -1,50 +1,24 @@
...
-    </div>
-  </div>
-  <div *ngSwitchCase="'notApplicable'">
-    <div class="container-fluid content padding-top-15 internal-page not-applicable-border">
-      <div class="col-xs-12 col-sm-8 col-sm-offset-2 col-md-6 col-md-offset-3 feature-content">
-        <div class="col-xs-12">
-          <i class="fa fa-flask fa-5x"></i>
-          <h2>Ooops</h2>
-          <p>This feature has been disabled by your admin page.</p>
+<div *ngIf="featureWarning" class="container-fluid container-cards-pf">
```
Comment (rohitkrai03) [inline projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.html:1]: @divyanshiGupta Is this backwards compatible with the way it was being used earlier with `FeatureFlagResolver`?

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.html (hunk 4 above)
Comment (divyanshiGupta) [reply projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.html:1]: Yes @rohitkrai03 this is backwards compatible, I have implemented this in a way that it doesn't breaks when used with `FeatureFlagResolver`. I have also tested it with a feature using `FeatureFlagResolver` at the route level.

File: projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts (hunk 3 above)
Comment (divyanshiGupta) [reply projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts:1]: @rohitkrai03 this is a duplicate service because at this point I didnt find any better and less time consuming way of doing this. You are right about the cyclic dependency issue and hence I tried to do it by creating the service interface in ngx-feature-flag and implementing it in fabric8-ui the way PipelineService is done in ngx-launcher but this also didnt worked out for me as some other issues were coming up. I agree with you that this can be solved once we get all the services in the SDK and we can use this service both in fabric8-ui and ngx-feature-flag from the SDK.

File: projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.html (hunk 4 above)
Comment (rohitkrai03) [reply projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.html:1]: That's good. Nice work there.

File: projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts (hunk 3 above)
Comment (rohitkrai03) [reply projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts:1]: @divyanshiGupta Understood. I'll add this use case in SDK doc.

**Review by rohitkrai03 — APPROVED** [review]
Tried it with the demo app. Works fine. Code looks good to me.

File: projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.ts [hunk 5]
```diff
@@ -1,29 +1,65 @@
...
-  private authService: AuthenticationService) {
-    if (authService.isLoggedIn()) {
-      this.userSubscription = userService.loggedInUser.subscribe(val => {
+  private featureWarningDataMap: Map<string, featureWarningData> = new Map([
+    [
+      'internal',
+      {
+        title: 'Internal',
+        description:
+          'These features are only available to Red Hat users and have no guarantee of performance or stability.Use these at your own risk.'
```
Comment (christianvogt) [inline projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.ts]: Missing a space after the period. Same for the other 2 messages.

File: projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.ts [hunk 6]
```diff
@@ -43,4 +79,42 @@ export class FeatureWarningPageComponent implements OnInit, OnDestroy {
...
+    const profile: ExtProfile = this.getTransientProfile();
+    this.enableFeatureService
+      .update(profile)
+      .pipe(first())
+      .subscribe(
+        (user: ExtUser): void => {
+          this.userService.currentLoggedInUser = user;
+          this.onOptInButtonClick.emit();
+          this.notifications.message({
+            message: `Feature Enabled!`,
```
Comment (christianvogt) [inline projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.ts]: Just a suggestion, but perhaps the message should also include the feature level that was enabled? Same for the error notification.

File: projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.less [hunk 7]
```diff
@@ -1,43 +1,31 @@
...
-          border-color: darken((@color-for-feature), 10%);
-        }
-      }
-    }
-  }
+.warning-page-container {
+  margin-top: 40%;
+  display: inline-block;
+  padding-left: 40px;
+  padding-right: 40px;
```
Comment (christianvogt) [inline projects/ngx-feature-flag/src/lib/warning-page/feature-warning-page.component.less]: Why are you setting `display: inline-block` and the `padding`? Removing them in the demo app leads to a better looking card on smaller devices.

File: projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+export class ExtProfile extends Profile {
+  contextInformation: any;
+  registrationCompleted: boolean;
+  featureLevel: string;
+}
+
+@Injectable()
+export class EnableFeatureService implements OnDestroy {
+  private headers = new HttpHeaders({ 'Content-Type': 'application/json' });
+  protected subscriptions: Subscription[] = [];
```
Comment (christianvogt) [inline projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts]: You don't need `subscriptions` as there are no subscriptions added in this file.

File: projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+  constructor(
+    protected auth: AuthenticationService,
+    protected http: HttpClient,
+    protected logger: Logger,
+    protected userService: UserService,
+    @Inject(FABRIC8_FEATURE_TOGGLES_API_URL) apiUrl: string
+  ) {
+    if (this.auth.getToken() != undefined) {
+      this.headers = this.headers.set('Authorization', 'Bearer ' + this.auth.getToken());
```
Comment (christianvogt) [inline projects/ngx-feature-flag/src/lib/service/enable-feature.service.ts]: We don't need this as the auth interceptor should take care of adding the `Authorization` header

File: projects/ngx-feature-flag/src/lib/feature-wrapper/feature-toggle.component.ts [hunk 10]
```diff
@@ -1,32 +1,57 @@
...
-  TemplateRef
-} from '@angular/core';
-import { Observable } from 'rxjs';
+import { Component, Input, OnInit, TemplateRef } from '@angular/core';
+import { Observable, of } from 'rxjs';
 import { FeatureTogglesService } from '../service/feature-toggles.service';
+import { Feature } from '../models/feature';
+import { map, catchError, tap } from 'rxjs/operators';
+
+export interface FeatureEnablementData {
```
Comment (christianvogt) [inline projects/ngx-feature-flag/src/lib/feature-wrapper/feature-toggle.component.ts]: This doesn't need to be exported.

**Review by christianvogt — APPROVED** [review]
